### DISCUSSION TITLE
AnonymousClassPatchPlugin For Kotlin

### DISCUSSION THREAD

**Discussion by wilgaboury** [HotswapProjects/HotswapAgent#547] [General]
I'm currently trying to use HotswapAgent for a kotlin project running on the JVM and while it works great for the most part, rearranging lambdas/anonymous classes in a file almost always causes class hierarchy errors. For normal Java hot swapping, I think this issue is handled by the `AnonymousClassPatchPlugin`. From looking at it's code, it relies on the assumption that anonymous classes will always have a name of the form `<class name>$<number>`, but this is not true in kotlin (which uses the format `<class name>$<method name>($<number>)+`. For my own purposes, I am planning to make a modified version of the plugin that will work for kotlin, but before I do, I wanted to ask here if this is an issue that has already been solved.

**Comment by skybber** [comment]
Hi, thanks for report. This problem is new for me and it has not been resolved in HotswapAgent yet, it would be great if you could share your soulution!

**Comment by wilgaboury** [comment]
Thanks. If I can get something working, I will certainly try and contribute it to this project.

**Comment by bric3** [comment]
> From looking at it's code, it relies on the assumption that anonymous classes will always have a name of the form `<class name>$<number>`, but this is not true in kotlin (which uses the format `<class name>$<method name>($<number>)+`. For my own purposes, I am planning to make a modified version of the plugin that will work for kotlin, but before I do, I wanted to ask here if this is an issue that has already been solved.

That feels similar to what I found there https://github.com/HotswapProjects/HotswapAgent/issues/420#issuecomment-1035675190, it's about `AnonymousClassInfos.lastModified`, I thought this was fixed.

Are you using the latest JBR and latest HA build.

**Comment by wilgaboury** [comment]
I was using the latest JBR and HA builds. I actually ended up migrating all my Kotlin code to Java for other reasons, so I haven't spent much time investigating this.

**Comment by Dustin4444** [comment]
> I'm currently trying to use HotswapAgent for a kotlin project running on the JVM and while it works great for the most part, rearranging lambdas/anonymous classes in a file almost always causes class hierarchy errors. For normal Java hot swapping, I think this issue is handled by the `AnonymousClassPatchPlugin`. From looking at it's code, it relies on the assumption that anonymous classes will always have a name of the form `<class name>$<number>`, but this is not true in kotlin (which uses the format `<class name>$<method name>($<number>)+`. For my own purposes, I am planning to make a modified version of the plugin that will work for kotlin, but before I do, I wanted to ask here if this is an issue that has already been solved.
